design/cassandra/sidecar.md

```diff
@@ -0,0 +1,77 @@
+## Sidecar Design Proposal
+
+### Consideration: REST API
+
+When thinking about how our sidecar will communicate with our controller, a natural solution that comes to mind is though a REST API. The sidecar will run an HTTP Server which the other party will call. This is the approach used by [Netflix's Priam](https://github.com/Netflix/priam/wiki).
+
+However, this includes some extra complications. Remember that, according to our goals we are designing a level-based system.
+First of all, some operations just take a long time. Backup, for example, might take hours to complete. 
+That means our operator must have an open TCP connection for all this time. If it gets interrupted, which we do expect to happen, this connection will be lost and we won't have any record that it ever happened. 
+
+
+This doesn't seem like the Kubernetes way of doing things. Consider this example, as one could think of the kubelet as a sidecar for Pods: 
+
+* **Question:** Does the scheduler ping the kubelet each time it schedules a Pod and then wait for an answer? 
+* **Answer:** No, it writes the `nodeName` field on the PodSpec. In other words, it writes a record of intent. No matter how many times the kubelet or the scheduler crashes it doesn't matter. The record of intent is there. 
+
+### Control-Loop Design
+
+* Based on our observations above, we design a method of communication in line with the Kubernetes philosophy.
+* When the controller wants to invoke a functionality in a Sidecar, it should write a record of intent in the Kubernetes API Objects (etcd). The sidecar will be watching the Kubernetes API and responding accordingly.
+* There are two approaches to represent the record of intent:
+  1. **Labels:**
+    * When the controller wants to communicate with a sidecar, it will write a predefined label in the Pod of the specific instance. For example, to communicate that we want an instance to terminate, we could write the label 'cassandra.rook.io/termination-timestamp=0`. The sidecar will see this and decommission the Cassandra instance. When it is done, it will change the label value to a valid timestamp. Then the controller will know to delete that instance.
+    * **Advantages:**
+      * Reuses Kubernetes built-in mechanisms
+      * Labels are query-able
+    * **Disadvantages:** 
+      * Doesn't support nested fields
+  2. **Member CRD**
+    * Each sidecar will watch an instance of a newly defined Member CRD and have its own `Spec` and `Status`.
+    * **Advantages:**
+      * More expressive and natural. Supports nested fields.
+      * Only our operator touches it. We don't expect it to happen often, but Pods are touched by pretty much everyone on the cluster. So if someone does stupid things, that affects us too.
+    * **Disadvantages:**
+      * Probably overkill to have for only a couple of fields.
+      * Induces an extra burden on etcd.
+
```

Comment:
I would prefer the first approach here, keeping it simple will probably make a lot of future things much simpler as well. Opening the config up for external messing seems like a great way for extra issues to be opened up. Also, nested fields don't really seem like a pressing requirement as far as I can see this. Worst case if we do happen to need such functionality is to use combinations of the fields.

With labels it pretty much a state machine which would be very easy to map out in our case in terms of flow of actions.